src/tags/object/Paragraphs/Phrases.js

```diff
@@ -20,9 +21,111 @@ const formatTime = (seconds) => {
+  const [animationKeyFrame, setAnimationKeyFrame] = useState(null);
+  const [seek, setSeek] = useState(0);
+  const [isSeek, setIsSeek] = useState(null);
   const cls = item.layoutClasses;
   const withAudio = !!item.audio;
+  let observer;
+
+  // default function to animate the reading line
+  const animateElement = useCallback(
+    (element, start, duration, isPlaying = true) => {
+      if (!element || !isFF(FF_LSDV_E_278)) return;
+
+      const _animationKeyFrame = element.animate(
+        [{ top: `${start}%` }, { top: '100%' }],
+        {
+          easing: 'linear',
+          duration: duration * 1000,
+        },
+      );
+
+      if (isPlaying)
+        _animationKeyFrame.play();
+      else
+        _animationKeyFrame.pause();
+
+      setAnimationKeyFrame(_animationKeyFrame);
+    },
+    [animationKeyFrame, setAnimationKeyFrame],
+  );
+
+  // this function is used to animate the reading line when user seek audio
+  const setSeekAnimation = useCallback(
+    (isSeeking) => {
+      if (!isFF(FF_LSDV_E_278)) return;
+      
+      const duration = item._value[playingId]?.duration || item._value[playingId]?.end - item._value[playingId]?.start;
+      const endTime = !item._value[playingId]?.end ? item._value[playingId]?.start + item._value[playingId]?.duration : item._value[playingId]?.end;
+      const seekDuration = endTime - seek.time;
+      const startValue = 100 - ((seekDuration * 100) / duration);
+
+      if (startValue > 0 && startValue < 100)
+        animateElement(activeRef.current?.querySelector('.reading-line'), startValue, seekDuration, seek.playing);
+      else
+        setIsSeek(isSeeking);
+    },
+    [seek, playingId],
+  );
+
+  // useRef to get the reading line element
+  const readingLineRef = useCallback(node => {
+    if(observer) {
+      observer.disconnect();
+    }
+
+    if(node !== null) {
+      const duration = item._value[playingId]?.duration || item._value[playingId]?.end - item._value[playingId]?.start;
+
+      if (!isNaN(duration)) {
+        animateElement(node, 0, duration, item.playing);
+      }
+
+      observer = new IntersectionObserver((entries) => {
+        setIsInViewport(entries[0].isIntersecting);
+      }, {
+        rootMargin: '0px',
+      });
+
+      observer.observe(node);
+    }
+  }, [playingId]);
+
+  useEffect(() => {
+    if (!isFF(FF_LSDV_E_278)) return;
+    
+    item.syncHandlers.set('seek', seek => {
+      item.handleSyncPlay(seek);
+      setSeek(seek);
+    });
+
+    return () => {
+      observer?.disconnect();
+    };
+  }, []);
+
+
+  // when user seek audio, the useEffect will be triggered and animate the reading line to the seek position
+  useEffect(() => {
+    setSeekAnimation(true);
+  }, [seek]);
+
+  // when user seek audio to a differente playing phrase, the useEffect will be triggered and animate the reading line to the seek position
+  useEffect(() => {
+    if (!isSeek) return;
```

Comment:
This value will be stale referenced based on the last value of playingId, is that the intention or should isSeek be added to dependencies list?

Reply:
That is the intention, if user seeks to a different playingId, the invisible bar should start from the place that user sought, so here when playingId changes I'm just checking if user sought to this, and if user sought, the animation will start from the correct place to be sync